### PR TITLE
API change mentioned in #4

### PR DESCRIPTION
**Pull request by masalachai**
Hi @nrempel ,

This is a tentative suggestion for refactoring the API as mentioned in issue #4. Currently just implemented for 1 operation (domain check), Let me know what you think.

There is a request type for the domain check operation called `epp_client::domain::check::Request` that just wraps the current request object `EppDomainCheck` and auto-implements an `EppRequest` trait through a derive macro named `EppRequest` that requires a response type to be set and does serialization and deserialization for the required types.

```rust
#[derive(EppRequest, Debug)]
#[response(EppDomainCheckResponse)]
pub struct Request(EppDomainCheck);
```

A var of this type can then be sent to the `transact_new` function (which can replace `transact()` in `client.rs` after we decide on an implementation).

Here the response type is the `EppDomainCheckResponse` alias for now (not sure if this also should be wrapped in a `domain::check::Response`). A call with this change looks something like this

```rust
let domains = vec!["eppdev.com", "eppdev.net"];
let request = domain::check::Request::new(domains, "some-cltrid");

let response = client.transact_new(&request).await.unwrap();
println!("{:?}", response);
```

### CODE REVIEW

**Comment by nrempel** [comment]
Nice, thanks for taking a run at this. This seems reasonable to me. What do you think @djc, any opinions here?

File: epp-client/src/tests/int.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use super::CLTRID;
```
Comment (djc) [inline epp-client/src/tests/int.rs:1]: Is there a reason for not putting this outside `src`, in the `tests` directory?

File: epp-client/src/epp/request.rs [hunk 2]
```diff
@@ -5,16 +5,32 @@ pub mod domain;
...
 };
-use crate::epp::xml::{EPP_CONTACT_XMLNS, EPP_DOMAIN_XMLNS, EPP_HOST_XMLNS, EPP_LANG, EPP_VERSION};
-use epp_client_macros::*;
+use crate::epp::xml::{
+    EppXml, EPP_CONTACT_XMLNS, EPP_DOMAIN_XMLNS, EPP_HOST_XMLNS, EPP_LANG, EPP_VERSION,
+};
+
+/// Trait to set correct value for xml tags when tags are being generated from generic types
+pub trait EppRequest {
+    type Output: EppXml + Debug;
```
Comment (djc) [inline epp-client/src/epp/request.rs:25]: Maybe name this `Response` instead of `Output`? Seems easier to follow in this case.

**Review by djc — COMMENTED** [review]
I really like how simple the new `Request` code looks! Some questions on how that is achieved:

File: epp-client-macros/src/lib.rs [hunk 3]
```diff
@@ -53,3 +53,59 @@ pub fn element_name_derive(input: TokenStream) -> TokenStream {
 
     element_name_macro(&ast)
 }
+
+fn epp_request_macro(ast: &syn::DeriveInput) -> proc_macro::TokenStream {
```
[the remark this answers is not in the document]
Comment (masalachai) [reply epp-client-macros/src/lib.rs:57]: Hi. There are a couple reasons why I opted for the macros. One is that the syntax is more declarative when defining the types and less verbose than having an impl block which would only have the associated type name be different for each implementation. 

The second reason is that many epp libs in other languages package extensions separately. I am undecided on whether we should do that also (package extensions in separate crates), but in the case that that happens, it's easier to use the macros for the common XML ser/de operations.

File: epp-client/src/tests/int.rs (hunk 1 above)
Comment (masalachai) [reply epp-client/src/tests/int.rs:1]: Yeah, it belongs in `tests/`. I'll move it.

File: epp-client/src/epp/request.rs (hunk 2 above)
Comment (masalachai) [reply epp-client/src/epp/request.rs:25]: Yes, that's more suitable. Thanks.

File: epp-client-macros/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (djc) [reply epp-client-macros/src/lib.rs:57]: Are you aware that trait methods can have default implementations attached to them?

[the remark this answers is not in the document]
Comment (masalachai) [reply epp-client-macros/src/lib.rs:57]: Yes, but the method to serialize the newtype `Request(EppDomainCheck)` references the field within the struct with `self.0` and that can't be used in a default implementation. And that method to serialize needs to be added to all the newtypes that wrap the `EppObject` types, so having a macro add them seems simpler.

Maybe there's another way to implement the `domain::check::Request` type.

**Comment by masalachai** [comment]
Hi @djc and @nrempel ,

I have reorganized some things in this PR to use the trait methods and work with extension types. The calls to `transact()` would look like the following with these changes

```rust
let domains = vec!["eppdev.com", "eppdev.net"];

// For a request without an extension
let request = domain::check::Request::<NoExtension>::new(domains, None, "some-cltrid");

// For a request with the NamestoreCheck extension
let ext = NamestoreCheck::new("dotCC");
let request = domain::check::Request::<NamestoreCheck>::new(domains, Some(ext), "some-cltrid");

let response = client.transact_new(&request).await.unwrap();
```

The `NoExtension` arg when there is no extension data to be added doesn't look very elegant, but I think it's easier to add extensions with this syntax with less boilerplate.

Let me know if this is something we can proceed with.

**Comment by djc** [comment]
I think we're getting closer! I would propose that `new()` does not take an extension argument (ditch the second argument), and we use `with_extension()` to always take an extension (so no `Option` wrapper). Both should then no longer need the explicit type annotation, since `new()` always parametrizes with `NoExtension` and `with_extension()` always passes in an argument of the required type.

It was quite intentional on my part that in my previous proposal the transaction ID does not become part of the `Request` itself and only is passed to the `transact_new()` call separately, since that seems to more closely reflect the data on the wire (and it also makes more conceptual sense to me).

**Comment by masalachai** [comment]
Hi @djc ,

Thanks for the feedback. @nrempel has implemented adding the extension data via a `with_extension()` while keeping the `trID` as an arg in the call to `transact_new()` in #21. It looks good to me. Please have a look and let us know if that approach is agreeable to you also.

**Comment by djc** [comment]
I worked with Nick on this this week, so I'm on board. Thanks for your prompt feedback throughout the process!

**Pull request closed without merging**